From 403bfc07c452c18690e5d8228f16fea3f0f84fbd Mon Sep 17 00:00:00 2001
From: Valerio Valerio <valerio.valerio@jollamobile.com>
Date: Wed, 10 Sep 2014 14:36:13 +0300
Subject: [PATCH 06/17] Add keepalive timer to IMAP IDLE service

This commit introduce a dependency on nemo-keepalive via
DEFINES+=USE_KEEPALIVE
---
 src/plugins/messageservices/imap/imap.pro        |  4 ++
 src/plugins/messageservices/imap/imapclient.cpp  |  3 ++
 src/plugins/messageservices/imap/imapclient.h    |  6 +++
 src/plugins/messageservices/imap/imapservice.cpp | 49 ++++++++++++++++++++++++
 src/plugins/messageservices/imap/imapservice.h   |  9 ++++-
 5 files changed, 70 insertions(+), 1 deletion(-)

diff --git a/src/plugins/messageservices/imap/imap.pro b/src/plugins/messageservices/imap/imap.pro
index 1c0ae6b..1b5742b 100644
--- a/src/plugins/messageservices/imap/imap.pro
+++ b/src/plugins/messageservices/imap/imap.pro
@@ -5,6 +5,10 @@ load(qt_plugin)
 
 QT = core network qmfclient qmfclient-private qmfmessageserver
 
+contains(DEFINES, USE_KEEPALIVE) {
+    PKGCONFIG += keepalive
+}
+
 contains(DEFINES,QT_QMF_USE_ALIGNEDTIMER) {
     QT += alignedtimer
 }
diff --git a/src/plugins/messageservices/imap/imapservice.cpp b/src/plugins/messageservices/imap/imapservice.cpp
index e926f1c..392aee1 100644
--- a/src/plugins/messageservices/imap/imapservice.cpp
+++ b/src/plugins/messageservices/imap/imapservice.cpp
@@ -1517,6 +1517,12 @@ ImapService::ImapService(const QMailAccountId &accountId)
       _networkSession(0),
       _networkSessionTimer(new QTimer(this))
 {
+#ifdef USE_KEEPALIVE
+    _backgroundActivity = new BackgroundActivity(this);
+    _backgroundActivity->setWakeupFrequency(BackgroundActivity::ThirtySeconds);
+    connect(_backgroundActivity, SIGNAL(running()), this, SLOT(startStopBackgroundActivity()));
+#endif
+
     QMailAccount account(accountId);
     if (!(account.status() & QMailAccount::CanSearchOnServer)) {
         account.setStatus(QMailAccount::CanSearchOnServer, true);
@@ -1978,8 +2014,25 @@ void ImapService::setPersistentConnectionStatus(bool status)
         }
     }
     _idling = status;
+#ifdef USE_KEEPALIVE
+    startStopBackgroundActivity();
+#endif
 }
 
+#ifdef USE_KEEPALIVE
+void ImapService::startStopBackgroundActivity()
+{
+    // start timer again if still in idle mode
+    if (_idling) {
+        qMailLog(Messaging) << Q_FUNC_INFO <<  "Starting keepalive";
+        _backgroundActivity->wait();
+    } else if (_backgroundActivity->isRunning()){
+        qMailLog(Messaging) << Q_FUNC_INFO << "Stopping keepalive";
+        _backgroundActivity->stop();
+    }
+}
+#endif
+
 class ImapConfigurator : public QMailMessageServiceConfigurator
 {
 public:
diff --git a/src/plugins/messageservices/imap/imapservice.h b/src/plugins/messageservices/imap/imapservice.h
index c9d38fe..21797de 100644
--- a/src/plugins/messageservices/imap/imapservice.h
+++ b/src/plugins/messageservices/imap/imapservice.h
@@ -51,6 +51,10 @@
 #include "imapclient.h"
 #include <qmailmessageservice.h>
 #include <QNetworkSession>
+
+#ifdef USE_KEEPALIVE
+#include <keepalive/backgroundactivity.h>
+#endif
 
 QT_BEGIN_NAMESPACE
 
@@ -71,6 +71,10 @@ protected slots:
     void openIdleSession();
     void closeIdleSession();
 
+#ifdef USE_KEEPALIVE
+    void startStopBackgroundActivity();
+#endif
+
 private slots:
     void onOnlineStateChanged(bool isOnline);
     void onSessionOpened();
@@ -110,6 +114,9 @@ private:
     QNetworkConfigurationManager    *_networkConfigManager;    // Qt network configuration manager
     QNetworkSession                 *_networkSession;          // Qt network session
     QTimer                          *_networkSessionTimer;
+#ifdef USE_KEEPALIVE
+    BackgroundActivity* _backgroundActivity;
+#endif
 };
 
 class ImapServicePlugin : public QMailMessageServicePlugin
-- 
2.7.4

